include/proxy-wasm/context.h

```diff
@@ -0,0 +1,649 @@
+
+  /**
+   * Call on a Root Context when a plugin is configured or reconfigured dynamically.
+   * @param plugin is the plugin which is being configured or reconfigured.
+   */
+  virtual bool onConfigure(std::shared_ptr<PluginBase> plugin);
+
+  // Context
+
+  /**
+   * Call on a host Context to create a corresponding Context in the VM.  Note:
+   * onNetworkNewConnection and onRequestHeaders() call onCreate().
+   * @param parent_context_id is the parent Context id for the context being created.  For a Network
+   * stream Context, this will be a Root Context id.
+   */
+  virtual void onCreate(uint32_t parent_context_id);
+
+  /**
+   * Network - see network_interface.h
+   */
+  ProxyAction onNetworkNewConnection() override;
+  ProxyAction onDownstreamData(uint32_t data_length, bool end_of_stream) override;
+  ProxyAction onUpstreamData(uint32_t data_length, bool end_of_stream) override;
+  using CloseType = NetworkInterface::CloseType;
+  void onDownstreamConnectionClose(CloseType close_type) override;
+  void onUpstreamConnectionClose(CloseType close_type) override;
+
+  /*
+   * HTTP - see http_interface.h
+   */
+  ProxyAction onRequestHeaders() override;
+  ProxyAction onRequestBody(uint32_t body_buffer_length, bool end_of_stream) override;
+  ProxyAction onRequestTrailers() override;
+  ProxyAction onRequestMetadata() override;
+  ProxyAction onResponseHeaders() override;
+  ProxyAction onResponseBody(uint32_t body_buffer_length, bool end_of_stream) override;
+  ProxyAction onResponseTrailers() override;
+  ProxyAction onResponseMetadata() override;
+
+  // Root Context async events
+
+  /**
+   * Called on a Root Context when a response arrives for an outstanding httpCall().
+   * @param token is the token returned by the corresponding httpCall().
+   * @param headers are the number of headers in the response.
+   * @param body_size is the size in bytes of the response body.
+   * @param trailers is the number of trailers in the response.
+   */
+  virtual void onHttpCallResponse(uint32_t token, uint32_t headers, uint32_t body_size,
+                                  uint32_t trailers);
+
+  /**
+   * Called on a Root Context to indicate that an Inter-VM shared queue message has arrived.
+   * @token is the token returned by registerSharedQueue().
+   */
+  virtual void onQueueReady(uint32_t token);
+
+  // Context
+
+  /**
+   * Call when a stream has completed (both sides have closed) or on a Root Context when the VM is
+   * shutting down.
+   * @return true for stream contexts or for Root Context(s) if the VM can shutdown, false for Root
+   * Context(s) if the VM should wait until the Root Context calls the proxy_done() ABI call.  Note:
+   * the VM may (optionally) shutdown after some configured timeout even if the Root Context does
+   * not call proxy_done().
+   */
+  virtual bool onDone();
+
+  /**
+   * Call when a stream should log the final status of the stream. Occurs after onDone().
+   */
+  void onLog() override;
```

Comment:
As mentioned elsewhere, this seems too imperative. Also, why can't we log in `onDone`?